### PR TITLE
JBMC: Add symex-based program loading

### PR DESCRIPTION
**Pull request by smowton**
This adds the final few pieces required and enables symex-driven program loading in JBMC, controlled by the `--symex-driven-lazy-loading` command-line option. In this mode, functions are effectively finalised one-by-one, including committing to final location numbers, loop numbers and so on, so that references to instructions taken by symex remain valid in the eyes of subsequent passes.

Some JBMC options are currently incompatible: some for fundamental reasons (`--lazy-loading` specifies an alternate, incompatible loading scheme; `--all-properties` requests a list of assertions with no particular entry-point in mind and therefore nowhere to symex from; `--full-slice` fundamentally requires a whole program available), and one that will require a little extra work (`java-unwind-static-init` is currently implemented in a way that clashes with symex-driven loading, but can be reworked in a subsequent PR).

The whole JBMC test corpus is used to test symex-driven loading, with a small number of tests tagged to indicate incompatibility.

### CODE REVIEW

**Comment by smowton** [comment]
@NathanJPhillips please review

**Comment by tautschnig** [comment]
Is it show-properties rather than all-properties that’s in conflict?

File: src/goto-programs/remove_exceptions.cpp
Comment (NathanJPhillips) [inline src/goto-programs/remove_exceptions.cpp:573]: Can the linter be fixed or this lint error ignored in review?

File: src/cbmc/bmc.cpp
Comment (NathanJPhillips) [inline src/cbmc/bmc.cpp]: This parameter needs to be passed by r-value if you're going to `std::move` it below. i.e. the model needs to be moved into this function, indicating to the caller that it is no longer valid after this function has been called.

File: src/jbmc/jbmc_parse_options.h
Comment (NathanJPhillips) [inline src/jbmc/jbmc_parse_options.h]: Is this normally indented?

**Review by NathanJPhillips — CHANGES_REQUESTED** [review]
Most significant request is not to std::move something passed into a function by l-value.

File: src/goto-programs/remove_exceptions.cpp
Comment (NathanJPhillips) [reply src/goto-programs/remove_exceptions.cpp:573]: Couple of other instances of this too.

File: regression/test.pl
Comment (tautschnig) [inline regression/test.pl]: 1. I think this commit (extending test.pl) could be moved to a separate PR
2. Actual change request: please extend the documentation that describes the syntax of test.desc files to cover the new tagging mechansim.

File: src/goto-symex/goto_symex_state.h
Comment (tautschnig) [inline src/goto-symex/goto_symex_state.h]: To me, this is the wrong place: this belong's to the state of the symbolic executor (`goto_symext`) and is not part of the symbolic program state.

File: src/cbmc/bmc.cpp
Comment (tautschnig) [inline src/cbmc/bmc.cpp]: Why did you need to rewrite the entire function when seemingly just the type of the parameter should change?

File: src/goto-symex/symex_assign.cpp
Comment (tautschnig) [inline src/goto-symex/symex_assign.cpp]: How come? This should have been removed by `remove_java_new`!

File: regression/jbmc-strings/java_char_array_init/test.desc
Comment (tautschnig) [inline regression/jbmc-strings/java_char_array_init/test.desc]: Please debug properly, for example by comparing the results of `--program-only`.

File: regression/strings-smoke-tests/java_append_char/test.desc
Comment (tautschnig) [inline regression/strings-smoke-tests/java_append_char/test.desc]: As above: please debug. No throwing-over-the-fence, please.

**Review by tautschnig — CHANGES_REQUESTED** [review]
"Symex: handle java-new side-effect" seems to be a wrong fix for whatever-the-symptom as `ID_java_new` ought not to exist at this point. Also, the tests need to be debugged fully. Otherwise a few comments.

File: src/goto-symex/symex_assign.cpp
Comment (smowton) [reply src/goto-symex/symex_assign.cpp]: Previously remove_java_new could not be run per-function, because it lowered into a sized malloc, and types could later gain fields when method conversion encountered a field access against a stubbed type. However this is no longer true, so the whole remove_java_new malarkey can now go away again. I'll submit a PR to that effect on Monday.

File: src/analyses/goto_check.h
Comment (martin-cs) [inline src/analyses/goto_check.h]: A string called "mode" is a bit leading ... no chance of a doxygen comment on what the inputs are is there?

File: src/goto-instrument/cover.h
Comment (martin-cs) [inline src/goto-instrument/cover.h]: I guess this fits with @peterschrammel 's changes to coverage...

**Review by martin-cs — APPROVED** [review]
Looks fine from the point of view of the bits of code I look after, hence approve.  BUT there are some pretty wide API-breaking changes here so I'm hoping that someone with a wider view of that part of the code can comment.

File: src/goto-programs/remove_exceptions.cpp
Comment (smowton) [reply src/goto-programs/remove_exceptions.cpp:573]: Long-standing problem; haven't the time or inclination to dig into the linter just now...

File: src/cbmc/bmc.cpp
Comment (smowton) [reply src/cbmc/bmc.cpp]: I needed to provide two alternate versions of `execute` (see below this function), which have different prefixes depending on whether they're using a lazy or an eager goto model, but then converge after symex. This particular factoring was to avoid duplicating the error-handling code between the two cases.

File: regression/jbmc-strings/java_char_array_init/test.desc
Comment (smowton) [reply regression/jbmc-strings/java_char_array_init/test.desc]: These two have now been solved and therefore this change dropped.

File: regression/strings-smoke-tests/java_append_char/test.desc
Comment (smowton) [reply regression/strings-smoke-tests/java_append_char/test.desc]: Change also dropped.

**Comment by smowton** [comment]
@tautschnig @NathanJPhillips changes applied, except as specifically noted. @tautschnig the dirty analysis check-if-function-already-analysed code has gone into `incremental_dirtyt`, defined alongside `dirtyt`.

@peterschrammel @thk123 @chrisr-diffblue this is now rebased and ready for review.

File: src/goto-symex/symex_main.cpp
Comment (tautschnig) [inline src/goto-symex/symex_main.cpp]: This ought to be an invariant. See https://github.com/diffblue/cbmc/pull/1799 for a discussion.

**Comment by smowton** [comment]
test.pl improvement moved to https://github.com/diffblue/cbmc/pull/1821 (but also still in this PR until that one lands).

File: src/goto-symex/symex_main.cpp
Comment (tautschnig) [inline src/goto-symex/symex_main.cpp]: I don't think that's needed here, let alone as a doxygen comment

File: src/goto-symex/symex_main.cpp
Comment (tautschnig) [inline src/goto-symex/symex_main.cpp]: I'd suggest to turn this into a `PRECONDITION` - the problem can be detected much earlier and should detected by the caller. (I know this isn't really your bug, but since lots of this code is being touched here ... if you don't mind?)

File: regression/test.pl
Comment (tautschnig) [reply regression/test.pl]: Ignore the first one, but the second item still stands.

Comment (smowton) [reply regression/test.pl]: Done in the other PR; will drop this commit here.

**Comment by smowton** [comment]
@tautschnig @karkhaz rebased this on top of the recent symex changes. Will peel off the first couple of commits as a separate PR later tonight.

File: src/goto-symex/symex_main.cpp
Comment (smowton) [reply src/goto-symex/symex_main.cpp]: Suggesting merging https://github.com/diffblue/cbmc/pull/1843 then it will be

**Comment by smowton** [comment]
This has been rebased to exclude the commits that landed in #1844. I'm now expecting @karkhaz to rebase #1641, then this PR will be next in line after some rewriting of its changes to the BMC driver (cbmc/bmc.cpp).

File: src/jbmc/jbmc_parse_options.cpp
Comment (NathanJPhillips) [inline src/jbmc/jbmc_parse_options.cpp]: This may abort without flushing JSON output.

File: src/jbmc/jbmc_parse_options.cpp
Comment (NathanJPhillips) [inline src/jbmc/jbmc_parse_options.cpp]: Where has this code come from?

**Comment by smowton** [comment]
@tautschnig @peterschrammel @thk123 @mgudemann this is now rebased and ready for review.

File: src/java_bytecode/java_bytecode_language.cpp
Comment (tautschnig) [inline src/java_bytecode/java_bytecode_language.cpp:817]: What is a "synthetic method"? Does this in any way relate to stubbing? If so, could we please use a consistent terminology across the code base?

File: src/java_bytecode/java_bytecode_language.cpp
Comment (tautschnig) [inline src/java_bytecode/java_bytecode_language.cpp:61]: Somewhere there must be `--help` output describing this, which also needs to be fixed. And so does the parse-options header file.

File: src/cbmc/bmc.cpp
Comment (tautschnig) [inline src/cbmc/bmc.cpp]: The examples seem strange: why should dumping a symbol table or goto functions require running symex? show-vcc of course does.

File: regression/jbmc-strings/StringIndexOf/test.desc
Comment (tautschnig) [inline regression/jbmc-strings/StringIndexOf/test.desc]: Doesn't @romainbrenguier have some fix for indexOf? Is this maybe no longer relevant?

File: src/java_bytecode/java_bytecode_language.cpp
Comment (thk123) [reply src/java_bytecode/java_bytecode_language.cpp:817]: I think synthetic methods are things like `clinit_wrapper` that we introduce to handle some JVM mechanisms, I could be wrong though? Perhaps could be defined in, e.g., `synthetic_methods_map.h`?

File: src/cbmc/bmc.cpp
Comment (thk123) [reply src/cbmc/bmc.cpp]: Because we use symex to load the functions, running show-goto-functions before symex would result in just `__CPROVER_Start`. The comment should probably make it explicit that this is used when looking at the "symbol-table... _in conjunction with symex driven loading_"

File: src/cbmc/bmc.h
Comment (thk123) [inline src/cbmc/bmc.h:148]: Add doc for the `callback_after_symex` specifically what the return value of the callback should indicate

File: src/java_bytecode/java_bytecode_language.cpp
Comment (smowton) [reply src/java_bytecode/java_bytecode_language.cpp:817]: Sort of. The synthetic methods are either clinit methods, generated for stub types, or static initializer method wrappers, generated for all static initializers to ensure they only run once, generated whether a stub type or not. In future the methods defined for array types (e.g. `.clone()`) will probably go in this map too. Stub bodies are currently provided with a different mechanism but could use this too.

In general a synthetic method is one generated internally by the language frontend without corresponding user bytecode. Stubs are thus synthetic, but not all synthetics are stubs.

**Comment by smowton** [comment]
TG bump: https://github.com/diffblue/test-gen/pull/1609

File: src/jbmc/jbmc_parse_options.cpp
Comment (thk123) [inline src/jbmc/jbmc_parse_options.cpp:384]: `for(const std::string &opt :`)?

File: src/java_bytecode/java_bytecode_language.cpp
Comment (smowton) [reply src/java_bytecode/java_bytecode_language.cpp:61]: Actually no, the option was added ages ago in anticipation of a driver program using it. That is now happening, but the anticipated name isn't the best.

File: src/cbmc/bmc.cpp
Comment (smowton) [reply src/cbmc/bmc.cpp]: Because with symex-driven lazy loading in play, symex is the only way functions get loaded. Thus show-goto-functions has to come after symex in that case.

File: src/jbmc/jbmc_parse_options.cpp
Comment (thk123) [inline src/jbmc/jbmc_parse_options.cpp:576]: Is this always null here? No - when `java-unwind-enum-static` is set? Perhaps a comment (though I'm wary it risks going out of date).

File: src/jbmc/jbmc_parse_options.cpp
Comment (thk123) [inline src/jbmc/jbmc_parse_options.cpp:1021]: This should be defined in a header file to allow re-use across different driver programs

File: regression/cbmc-java/NondetInit/test_lazy.desc
Comment (thk123) [inline regression/cbmc-java/NondetInit/test_lazy.desc:1]: Probably good to add a comment to this `test.desc` explaining why this is an expected failure

Comment (thk123) [reply regression/cbmc-java/NondetInit/test_lazy.desc:1]: I see it is because of the `lazy-methods` flag. I still think they'd benefit from a uniform comment, but you may disagree.

File: regression/cbmc-java/lvt-groovy/test.desc
Comment (thk123) [inline regression/cbmc-java/lvt-groovy/test.desc]: Probably should call it "symex driven lazy loading" since otherwise will get confusing.

File: regression/cbmc-java/lvt-unexpected/test.desc
Comment (thk123) [inline regression/cbmc-java/lvt-unexpected/test.desc]: Probably should call it "symex driven lazy loading" since otherwise will get confusing. et a.

File: regression/strings-smoke-tests/CMakeLists.txt
Comment (thk123) [inline regression/strings-smoke-tests/CMakeLists.txt:5]: Should these profiles be somehow added to Makefiles as well? Is that even feasible?

**Review by thk123 — COMMENTED** [review]
Nothing blocking - some places where I think a few things could be clearer.

File: regression/cbmc-java/NondetInit/test_lazy.desc
Comment (allredj) [reply regression/cbmc-java/NondetInit/test_lazy.desc:1]: +1 for having a comment on each of the skipped desc to mention the reason they are skipped.

File: regression/jbmc-strings/StringIndexOf/test.desc
Comment (smowton) [reply regression/jbmc-strings/StringIndexOf/test.desc]: Re-tested and this indeed appears to have cleared in the meantime. Removed the tag.

**Comment by smowton** [comment]
@tautschnig comments addressed

File: src/java_bytecode/java_bytecode_language.cpp
Comment (thk123) [reply src/java_bytecode/java_bytecode_language.cpp:817]: I suggest this definition goes somewhere find able in the codebase.

Comment (smowton) [reply src/java_bytecode/java_bytecode_language.cpp:817]: This isn't really related to this PR, so I've made https://github.com/diffblue/cbmc/pull/1953

File: src/cbmc/bmc.cpp
Comment (smowton) [reply src/cbmc/bmc.cpp]: @thk123 added comment.

**Review by peterschrammel — COMMENTED** [review]
Re --show-properties: there is a --drop-unused-functions option that also works in combination with  --show-properties. Maybe you can achieve a similar behaviour for --show-properties when used together with your option.

File: src/jbmc/jbmc_parse_options.cpp
Comment (smowton) [reply src/jbmc/jbmc_parse_options.cpp:576]: Comment added

File: src/jbmc/jbmc_parse_options.cpp
Comment (smowton) [reply src/jbmc/jbmc_parse_options.cpp:1021]: Hmmm, it is pretty jbmc-specific at the moment though. I can't see anything except the option name being shareable with test-gen at the moment -- and it surely doesn't belong to bmct or goto_symext, both of which don't care exactly how their get_goto_program function works.

In short I agree some sort of factoring is in order at some point, but I can't see how to do it just now.

File: src/java_bytecode/java_bytecode_language.cpp
Comment (tautschnig) [reply src/java_bytecode/java_bytecode_language.cpp:61]: So you're removing an option? Different kind of update, but still one expected?!

File: src/cbmc/bmc.cpp
Comment (tautschnig) [reply src/cbmc/bmc.cpp]: Sorry, but having to do symbolic execution in order to obtain syntactic information makes no sense. It's ok to forbid lazy loading together with show-goto-functions, but please don't produce a tool that requires a semantic analysis to produce syntactic information.

File: src/jbmc/jbmc_parse_options.cpp
Comment (tautschnig) [reply src/jbmc/jbmc_parse_options.cpp:1021]: But maybe in future CBMC will care?

File: src/java_bytecode/java_bytecode_language.cpp
Comment (smowton) [reply src/java_bytecode/java_bytecode_language.cpp:61]: Nobody (except this file) ever accepted it until now. JBMC now accepts "symex-driven-lazy-loading".

Comment (tautschnig) [reply src/java_bytecode/java_bytecode_language.cpp:61]: Ok.

File: src/cbmc/bmc.cpp
Comment (smowton) [reply src/cbmc/bmc.cpp]: This is a reasonable thing to want for debug purposes: if I want to know what got loaded by symex-driven loading, I would pass `--symex-driven-lazy-loading --list-goto-functions` for example. If I want the whole class file I just don't use `symex-driven-lazy-loading`, which processes show/list-goto-functions the old fashioned way.

**Comment by smowton** [comment]
@thk123 @tautschnig all comments addressed, please re-review

File: regression/strings-smoke-tests/CMakeLists.txt
Comment (smowton) [reply regression/strings-smoke-tests/CMakeLists.txt:5]: Added

File: src/jbmc/jbmc_parse_options.cpp
Comment (smowton) [reply src/jbmc/jbmc_parse_options.cpp:1021]: This is possible. For now CBMC uses none of the lazy loading infrastructure at all, and I believe @kroening's plan is for it to stay that way for a while. If and when it does, some factoring will be in order.

File: src/cbmc/bmc.cpp
Comment (tautschnig) [reply src/cbmc/bmc.cpp]: This implies that the behaviour of show-goto-functions is context-dependent. My take on this is: if you want a way to show the subset of functions that have been loaded during symbolic execution, add a new command-line option. Don't overload an existing one.

Comment (smowton) [reply src/cbmc/bmc.cpp]: We already have `--lazy-methods`, under which `--show-goto-functions` will again show whatever was loaded, not all the functions in a particular class file.

Comment (tautschnig) [reply src/cbmc/bmc.cpp]: It didn't claim that was good either ;-)

What I would suggest is additional text to the help output that explains the interaction of the command-line options.

**Review by thk123 — APPROVED** [review]
LGTM (TG bump passes as well).

FWIW my two cents re: `--show-goto-functions` is that the interaction is expected and useful when combined with `--symex-driven-lazy-loading`, it has certainly never surprised me that that is how it behaves for `--lazy-methods`.

I would be disappointed if we lost the ability to show loaded GOTO functions and it seems unnecessarily confusing to have a `--show-loaded-goto-functions` that is only useful when you've got a lazy loading on, and `--show-goto-functions` that's only useful when you've got it off.

**Comment by smowton** [comment]
@peterschrammel after a little more thought I realised show-properties could be accommodated similarly to --show-symbol-table et al. I've added a commit switching it back on, along with its test.

File: src/cbmc/bmc.cpp
Comment (smowton) [reply src/cbmc/bmc.cpp]: https://github.com/diffblue/cbmc/pull/1957

File: src/cbmc/bmc.cpp
Comment (NathanJPhillips) [inline src/cbmc/bmc.cpp]: Is this done by a front-end or a driver? If the former then hyphenate, if the latter then reword here and rename the variable.

File: src/jbmc/jbmc_parse_options.cpp
Comment (NathanJPhillips) [reply src/jbmc/jbmc_parse_options.cpp:384]: They look like `const char *`'s to me ;-)
If you want to do the construction of a `std::string` for some reason and you want to push it into the for loop then at least use `const std::string` instead of relying on lifetime-extension using `const std::string &` as this will indicate that this is newly created in this scope, not a reference to something created elsewhere.

**Review by NathanJPhillips — APPROVED** [review]
Might want to squash a couple of the later test commits.

**Comment by smowton** [comment]
@tautschnig are you satisfied, can we go ahead?

**Review by tautschnig — APPROVED** [review]

File: src/cbmc/bmc.cpp
Comment (smowton) [reply src/cbmc/bmc.cpp]: Good point, done

**Review by peterschrammel — APPROVED** [review]

**Review by chrisr-diffblue — APPROVED** [review]
Reviewing the parts I'm code owner for looks fine by me.

**Pull request merged**
